### PR TITLE
[inductor] Run symmetric matrix products together

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #194181
* #194180
* __->__ #194179
* #194178
* #194177

Human commentary from the author: "and add architecture descriptions and examples to each PR to make it clear"

> AI-assisted change description:
>
> ## Summary
>
> Add reusable execution plans for running many independent symmetric matrix
> products and Newton-Schulz updates. The plans cover singleton/sequential,
> homogeneous packed-batch, and heterogeneous grouped workloads while sharing
> compiled code, metadata, outputs, and workspaces across calls.
>
> ## Architecture
>
> The planner separates workload organization from the Newton-Schulz
> calculation:
>
> ```text
> list of update matrices
>          |
>          +-- 1-3 matrices --------> SequentialSymmetricMuonPlan
>          |
>          +-- same large shape ----> PackedSymmetricMuonPlan
>          |
>          +-- mixed compatible ----> GroupedMuonPlan
>                                          |
>                                          v
>                               GroupedSymmetricPlan
>                               - fixed output/workspace tensors
>                               - device metadata for shapes/strides/pointers
>                               - one grouped CuTeDSL launch
>                               - one mirror pass for the other triangle
> ```
>
> `SymmetricMuonPlan` owns the workspaces for one shape. The sequential plan
> reuses those shape-specific plans for small groups. The packed plan stacks
> same-shaped matrices into a homogeneous batch. The grouped plan decomposes
> heterogeneous matrices into row blocks and launches them through the
> existing grouped CuTeDSL GEMM kernel.
>
> Plans retain their input/output tensors and metadata because pointer setup
> and compilation are part of the overhead this change removes. Callers must
> reuse the returned tensors rather than assuming each invocation allocates a
> new functional result.
>
> ## Example
>
> ```python
> from torch._inductor.kernel.grouped_symmetric_mm import GroupedMuonPlan
>
> shapes = [(1536, 5120), (1536, 5120), (2048, 7168), (3200, 5120)]
> updates = [
>     torch.randn(shape, device="cuda", dtype=torch.bfloat16)
>     for shape in shapes
> ]
> plan = GroupedMuonPlan(
>     shapes,
>     updates[0].device,
>     coefficients=(3.4445, -4.775, 2.0315),
>     steps=5,
> )
> orthogonalized = plan(updates, eps=1e-7)
> ```
>
> This is an internal reusable-plan API. The following commits connect it to
> compiled `torch.optim.Muon` automatically.
>
> ## Benchmark results
>
> Homogeneous expert-batch measurements on an NVIDIA GB200 showed the
> parallelism recovered by planning matrices together:
>
> | batched matrix shape | batch counts | speedup over separate cuBLAS GEMMs |
> |---|---:|---:|
> | `[E, 2048, 7168]` | 4-384 | 1.62-1.75x |
> | `[E, 2048, 6144]` | 32-256 | 1.62-1.83x |
>
> A single `(2048, 7168)` matrix was only 1.04x faster, which motivated the
> sequential/packed/grouped plan split instead of forcing every workload
> through one grouped implementation. These are development microbenchmarks,
> not performance guarantees.
>
> ## Test Plan
>
> ```text
> CUDA_VISIBLE_DEVICES=0 TORCHINDUCTOR_CACHE_DIR=/tmp/codex-pytorch2-final-symmetric-cache .venv/bin/python test/inductor/test_symmetric_mm.py -v
> source .venv/bin/activate && lintrunner -a
> ```
>
> Authored with an AI assistant.

### CODE REVIEW

**Comment by mlazos** [comment]
## AI-Review

> *Automated multi-agent review (architecture, correctness, simplification, generalization, and GPT-5.6 final approval). This is AI-generated — please verify findings manually.*

---

## Verified Issues
- **[high] Mutable list aliases can invalidate cached GPU pointers** — Location: `torch/_inductor/kernel/grouped_symmetric_mm.py:111-159`. Trigger: replace or append an entry in a caller-owned `inputs`, `c`, or `outputs` list after construction. Failure sequence: tensor addresses are cached separately while the mutable lists remain aliased; replacement can free the referenced storage, and appending outputs expands the launch grid beyond metadata lengths. Impact: stale or out-of-bounds GPU accesses, corruption, or illegal-memory-access failures. Reviewers: Correctness, Generalization.
- **[high] Tall Muon updates are returned transposed** — Location: `torch/_inductor/kernel/grouped_symmetric_mm.py:314-331, 391-415, 511-546`. Trigger: process an update whose first dimension exceeds its second. Failure sequence: the update is transposed into `(min, max)` form, but the selected output is returned without restoring its original orientation. Impact: downstream optimizer updates receive the wrong shape and fail. Reviewers: Architecture, Correctness, Generalization.
- **[high] Sequential plans cannot handle heterogeneous Gram sizes** — Location: `torch/_inductor/kernel/grouped_symmetric_mm.py:344-350`. Trigger: construct with shapes such as `(512, 1024)` and `(1024, 2048)`. Failure sequence: the first plan’s `512x512` workspace is passed to the second plan, whose validation requires `1024x1024`. Impact: construction raises, making sequential heterogeneous workloads unusable. Reviewers: Architecture, Correctness, Generalization.
- **[high] Module import fails in CPU-only import-coverage tests** — Location: `torch/_inductor/kernel/grouped_symmetric_mm.py:5-18`. Trigger: `test_modules_can_be_imported` runs where optional CUTLASS dependencies are absent. Failure sequence: module-scope CUTLASS imports raise `ModuleNotFoundError`, and this wrapper is not allowlisted. Impact: deterministic CI failure and an unimportable private module. Reviewers: Architecture.
- **[medium] `alpha` is ignored when `c` is absent** — Location: `torch/_inductor/kernel/grouped_symmetric_mm.py:43-53`. Trigger: construct `GroupedSymmetricPlan([x], alpha=2.0)` without `c`. Failure sequence: the grouped GEMM writes the unscaled product and the no-`C` mirror branch only copies its transpose. Impact: silently returns `x @ x.T` instead of `2 * (x @ x.T)`. Reviewers: Architecture, Correctness, Generalization.
- **[medium] Pre-Blackwell GPUs reach a Blackwell-only kernel** — Location: `torch/_inductor/kernel/grouped_symmetric_mm.py:81-82, 218-237`. Trigger: construct the plan with valid BF16 tensors on an A100 or H100. Failure sequence: validation rejects only HIP, then CuTeDSL compiles a `tcgen05` grouped kernel whose architecture check rejects the target. Impact: plan construction fails inside compilation. Reviewers: Architecture, Correctness, Generalization.
- **[low] Negative odd step counts return uninitialized storage** — Location: `torch/_inductor/kernel/grouped_symmetric_mm.py:306-307, 383-384, 509`. Trigger: use `steps=-1`, which existing Muon accepts. Failure sequence: `range(-1)` performs no iterations, but odd parity selects the untouched alternate buffer. Impact: arbitrary BF16 data is returned instead of the normalized input. Reviewers: Correctness, Generalization.
- **[low] One-CTA mode is incompatible with its default tile** — Location: `torch/_inductor/kernel/grouped_symmetric_mm.py:72-74, 218-224`. Trigger: specify `use_2cta=False` without overriding `mma_tiler`. Failure sequence: tile-M remains 256, while the one-CTA BF16 kernel supports only 64 or 128. Impact: kernel setup or compilation fails on an otherwise supported GPU. Reviewers: Correctness.
